File: java/P0234.java
class P0234{
    public static void main(String[] args) {
        
    }
    static class ListNode {
     int val;
     ListNode next;
     ListNode(int x) { val = x; }
  }
 
    static class Solution {
        public boolean isPalindrome(ListNode head) {
            if(head==null || head.next==null)return true;
            Deque<ListNode> dq = new LinkedList<>();
            ListNode tail = head;
            int len = 0;
            while(tail!=null){
                len++;
                tail = tail.next;    
            }
            tail = head;
            int half = len/2;
            while(half-->0){
                dq.addLast(tail);
                tail = tail.next;
            }
            if(len%2==1)tail = tail.next;
            while(tail!=null){
                if(tail.val==dq.removeLast().val)tail = tail.next;
                else return false;    
            }
            return true;
        }
    }
}

File: go/P0234.go
//题意：判断回文链表，要求时间On空间O1
//题解：快慢指针，到一半的时候反转后半段链表。
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func isPalindrome(head *ListNode) bool {
    if head == nil || head.Next == nil{
        return true
    }
    slow,fast := head,head
    for fast!=nil && fast.Next!=nil{
        fast = fast.Next.Next
        slow = slow.Next
    }
    var pre,now,nxt *ListNode
    now = slow
    for now!=nil{
        nxt = now.Next
        now.Next = pre
        pre = now
        now = nxt
    }
    beg := pre
    for head!=nil && beg!=nil && beg!=head{
        if head.Val!=beg.Val{
            return false
        }
        head = head.Next
        beg = beg.Next
    }
    return true
}